pkg/resources/storage_integration.go

```diff
@@ -118,284 +120,262 @@ func StorageIntegration() *schema.Resource {
+			Path: loc,
+		}
+	}
 
-	stmt.SetStringList("STORAGE_ALLOWED_LOCATIONS", expandStringList(d.Get("storage_allowed_locations").([]interface{})))
+	req := sdk.NewCreateStorageIntegrationRequest(name, enabled, storageAllowedLocations)
 
-	// Set optional fields
 	if v, ok := d.GetOk("comment"); ok {
-		stmt.SetString(`COMMENT`, v.(string))
+		req.WithComment(sdk.String(v.(string)))
 	}
 
 	if _, ok := d.GetOk("storage_blocked_locations"); ok {
-		stmt.SetStringList("STORAGE_BLOCKED_LOCATIONS", expandStringList(d.Get("storage_blocked_locations").([]interface{})))
+		stringStorageBlockedLocations := expandStringList(d.Get("storage_blocked_locations").([]any))
+		storageBlockedLocations := make([]sdk.StorageLocation, len(stringStorageBlockedLocations))
+		for i, loc := range stringStorageBlockedLocations {
+			storageBlockedLocations[i] = sdk.StorageLocation{
+				Path: loc,
+			}
+		}
 	}
 
-	if _, ok := d.GetOk("storage_aws_object_acl"); ok {
-		stmt.SetString("STORAGE_AWS_OBJECT_ACL", d.Get("storage_aws_object_acl").(string))
-	}
+	storageProvider := d.Get("storage_provider").(string)
 
-	// Now, set the storage provider
-	if err := setStorageProviderSettings(d, stmt); err != nil {
-		return err
+	switch storageProvider {
+	case "S3", "S3GOV":
+		v, ok := d.GetOk("storage_aws_role_arn")
+		if !ok {
+			return fmt.Errorf("if you use the S3 storage provider you must specify a storage_aws_role_arn")
+		}
+
+		s3Params := sdk.NewS3StorageParamsRequest(v.(string))
+		if _, ok := d.GetOk("storage_aws_object_acl"); ok {
+			s3Params.WithStorageAwsObjectAcl(sdk.String(d.Get("storage_aws_object_acl").(string)))
+		}
+		req.WithS3StorageProviderParams(s3Params)
+	case "AZURE":
+		v, ok := d.GetOk("azure_tenant_id")
+		if !ok {
+			return fmt.Errorf("if you use the Azure storage provider you must specify an azure_tenant_id")
+		}
+		req.WithAzureStorageProviderParams(sdk.NewAzureStorageParamsRequest(sdk.String(v.(string))))
+	case "GCS":
+		req.WithGCSStorageProviderParams(sdk.NewGCSStorageParamsRequest())
+	default:
+		return fmt.Errorf("unexpected provider %v", storageProvider)
 	}
-	if err := snowflake.Exec(db, stmt.Statement()); err != nil {
+
+	if err := client.StorageIntegrations.Create(ctx, req); err != nil {
 		return fmt.Errorf("error creating storage integration: %w", err)
 	}
 
-	d.SetId(name)
-
+	d.SetId(helpers.EncodeSnowflakeID(name))
 	return ReadStorageIntegration(d, meta)
 }
 
-// ReadStorageIntegration implements schema.ReadFunc.
-func ReadStorageIntegration(d *schema.ResourceData, meta interface{}) error {
+func ReadStorageIntegration(d *schema.ResourceData, meta any) error {
 	db := meta.(*sql.DB)
-	id := d.Id()
-
-	stmt := snowflake.NewStorageIntegrationBuilder(d.Id()).Show()
-	row := snowflake.QueryRow(db, stmt)
-
-	// Some properties can come from the SHOW INTEGRATION call
+	ctx := context.Background()
+	client := sdk.NewClientFromDB(db)
+	id, ok := helpers.DecodeSnowflakeID(d.Id()).(sdk.AccountObjectIdentifier)
+	if !ok {
```

Comment:
We are currently not doing a check like this in other places. I wonder if putting such an error is enough for the user. After all, id should be set by us, so we should have no trouble decoding it (if someone was not messing with state). We can leave it, but this will be bigger discussion with resources redesign.